From 152de8c68d13845592e8e511136842bcdb691063 Mon Sep 17 00:00:00 2001
From: Namjae Jeon <namjae.jeon@samsung.com>
Date: Sat, 29 May 2021 07:59:40 +0900
Subject: [PATCH] cifsd: fix Control flow issues in
 ksmbd_build_ntlmssp_challenge_blob()
Git-commit: 152de8c68d13845592e8e511136842bcdb691063
Patch-mainline: v5.15-rc1
References: bsn#12
Modified-by-SEL: No


Fix a defect reported by Coverity Scan.

*** CID 1504970:  Control flow issues  (NO_EFFECT)
/fs/cifsd/auth.c: 622 in ksmbd_build_ntlmssp_challenge_blob()
616             name = kmalloc(2 + UNICODE_LEN(len), GFP_KERNEL);
617             if (!name)
618                     return -ENOMEM;
619
620             conv_len = smb_strtoUTF16((__le16 *)name,
ksmbd_netbios_name(), len,
621                                       sess->conn->local_nls);
>>>     CID 1504970:  Control flow issues  (NO_EFFECT)
>>>     This less-than-zero comparison of an unsigned value is never true.
622             if (conv_len < 0 || conv_len > len) {
623                     kfree(name);
624                     return -EINVAL;
625             }
626
627             uni_len = UNICODE_LEN(conv_len);

Reported-by: Coverity Scan <scan-admin@coverity.com>
Signed-off-by: Namjae Jeon <namjae.jeon@samsung.com>
Signed-off-by: Steve French <stfrench@microsoft.com>
Signed-off-by: Guoqing Jiang <guoqing.jiang@suse.com>
---
 fs/cifsd/auth.c | 4 ++--
 1 file changed, 2 insertions(+), 2 deletions(-)

diff --git a/fs/cifsd/auth.c b/fs/cifsd/auth.c
index 9b86cf4fd73f..5f47de49c05d 100644
--- a/fs/cifsd/auth.c
+++ b/fs/cifsd/auth.c
@@ -584,8 +584,8 @@ ksmbd_build_ntlmssp_challenge_blob(struct challenge_message *chgblob,
 	struct target_info *tinfo;
 	wchar_t *name;
 	__u8 *target_name;
-	unsigned int len, flags, blob_off, blob_len, type, target_info_len = 0;
-	unsigned int uni_len, conv_len;
+	unsigned int flags, blob_off, blob_len, type, target_info_len = 0;
+	int len, uni_len, conv_len;
 	int cflags = sess->ntlmssp.client_flags;
 
 	memcpy(chgblob->Signature, NTLMSSP_SIGNATURE, 8);
-- 
2.33.0

